### DISCUSSION TITLE
[ crazyflie-lib-python] send_setpoint(roll,pitch,yawrate,thrust) question: how to convert from Newton(N) to the thrust value?

### DISCUSSION THREAD

**Discussion by tiagomsr99** [bitcraze/discussions#167] [Q&A] [answered]
Hi. I'm currently working on my master thesis. I'm developing a testbed/framework to test n-Crazyflie's indoor using the Lighthouse positioning system. So far I've developed a MATLAB/Simulink controller and some Python scripts to connect, get logs, etc. from the Crazyflie. Currently I can send the Crazyflie's logging data in real-time to my Simulink controller.

Right now I must send my controller's inputs and intend to use this, from the Python lib:
` send_setpoint(roll, pitch, yawrate, thrust)`

According to the online documentation this thrust value is an integer between 10000-60000. **How can convert my controllers thrust output, in Newtons, to this thrust value?**

Searching online I've found [this topic](https://forum.bitcraze.io/viewtopic.php?t=4434) at the old Bitcraze forum. Supposedly using this I can convert from thrust (10000-60000) to acceleration:
`CONTROL_TO_ACC (GRAVITY_MAGNITUDE*60.0f/CRAZYFLIE_WEIGHT_grams/65536.0f)`
But right now I'm confused on how to use this conversion.

Is this the correct use of the formula?
`acceleration (in m/s^2) = thrust (an integer) * CONTROL_TO_ACC`
This doesn't seem right for thrust values between 10000 and 60000

I've seen the mapping between thrust (in grams), PWM and RPM, (https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/pwm-to-thrust/) and according to this topic[this topic](https://forum.bitcraze.io/viewtopic.php?t=1391) this integer thrust is the mean of all motors's PWM.

But since I want to use send_setpoint(), to use PWM, battery or RPM related conversions seems too much. I feel like I'm missing something on the CONTROL_TO_ACC formula.

Thanks in advance.

**Comment by knmcguire** [comment, marked as answer]
Hi! I just changed the wiki link to the new link on our web in your post, since we are phasing out the wiki.

This line in the firmware:

https://github.com/bitcraze/crazyflie-firmware/blob/8c925aa7abd79ac05ef7d94b98bfacd4e37441fd/src/modules/src/estimator_kalman.c#L184

is related to the kalman filter, as it uses the control inputs to predict where the next position of the crazyflie will be. So it should be useable for what you want to do but it might not be super accurate either as we haven't tested if this fits the measurements... This filter was a contribution by ETH. They have done system identification a while ago on the Crazyflie 2.0 (which should hold for the 2.1), so that might be interesting for you to look at as well.

We are in the process of making a better test rig for these measurements, but currently it's summer here so that's on a hold. In 2 weeks @tobbeanton will be back so he can give an update about that then

**Reply by knmcguire** [reply]
Here it the thesis from eth I mentioned: https://www.research-collection.ethz.ch/handle/20.500.11850/214143

**Comment by tiagomsr99** [comment]
Thank you so much, I've managed to control a Crazyflie using a Matlab/Simulink controller! 

The controller works well but has significant static error. My controller has no integral effect (it's a Mellinger controller). I know that the CF's Mellinger controller has [an added integral term to compensate battery voltage drop, etc,](https://github.com/bitcraze/crazyflie-firmware/blob/master/src/modules/src/controller_mellinger.c#L32-L36) so I intend one as well on my Simulink controller. My university is currently closed for summer so I can't test this yet.

**Reply by knmcguire** [reply]
ah great! Let us know if it works out. 

Did the above proposed answer help out in your process? If so you can mark it as answer and if not, you can write your own 'suggest an answer' underneath here and mark that as answer.

**Reply by tiagomsr99** [reply]
Yeah it did! Sorry it's my first time using Github discussions.
